streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java

```diff
@@ -789,6 +806,158 @@ public void shouldWriteLatestOffsetsToCheckpointOnShutdown() throws Exception {
+                        if (recordMetadata.partition() == 0) {
+                            if (throwException.compareAndSet(true, false)) {
+                                throw new TaskCorruptedException(Collections.singleton(task00));
+                            }
+                            stateStore.put(record.key(), record.value());
+                        } else {
+                            stateStore.put(record.key(), record.value());
+                            if (restoredOffsetsForPartition0.get() > 0) {
+                                latch.countDown();
+                            }
+                        }
+                    });
+                }
+
+                @Override
+                public void close() {
+                    Processor.super.close();
+                }
+            }, "source")
+            .addStateStore(stateStore, "processor")
+            .addSink("sink", MULTI_PARTITION_OUTPUT_TOPIC, "processor");
+
+        final KafkaStreams kafkaStreams = new KafkaStreams(topology, streamsConfiguration);
+        kafkaStreams.setGlobalStateRestoreListener(new StateRestoreListener() {
+            @Override
+            public void onRestoreStart(final TopicPartition topicPartition,
+                                       final String storeName,
+                                       final long startingOffset,
+                                       final long endingOffset) {
+                if (topicPartition.partition() == 0) {
+                    System.out.println("Restore listener - Starting offset: " + startingOffset);
+                }
+            }
+            @Override
+            public void onBatchRestored(final TopicPartition topicPartition,
+                                        final String storeName,
+                                        final long batchEndOffset,
+                                        final long numRestored) {
+                if (topicPartition.partition() == 0) {
+                    System.out.println("Restore listener - Batch end offset: " + batchEndOffset);
+                    restoredOffsetsForPartition0.set(batchEndOffset);
+                }
+            }
+            @Override
+            public void onRestoreEnd(final TopicPartition topicPartition,
+                                     final String storeName,
+                                     final long totalRestored) {
+                if (topicPartition.partition() == 0) {
+                    System.out.println("Restore listener - Total restored: " + totalRestored);
+                }
+            }
+        });
+        startApplicationAndWaitUntilRunning(Collections.singletonList(kafkaStreams), Duration.ofSeconds(60));
+        ensureCommittedRecordsInTopic(
+            applicationId + "-" + stateStoreName + "-changelog",
+            2000,
+            StringDeserializer.class,
+            StringDeserializer.class
+        );
+        throwException.set(true);
```

Comment:
Is there a risk here that I may process all 30k keys before I manage to throw the exception?

Reply:
Why is that a risk?
If you are referring to the `else`-branch in the processor, I changed that in the last commit.

Reply:
OK, I see that there is a chance to not throw the exception in the `if`-branch of the processor. I added a second batch of records to add to the input topic after ensuring committed records in the changelog topic.